### ISSUE TITLE
[Replace] Document that replacements are not done before all other plugins/Rollup executes even when putting the `replace` plugin first.

### ISSUE THREAD

**Issue by MelvinWM** [rollup/plugins#910]
- Rollup Plugin Name: @rollup/plugin-replace
- Rollup Plugin Version: 2.4.2

Documentation Is:

- [x] Missing
- [x] Needed
- [ ] Confusing
- [x] Not Sure?

### Please Explain in Detail...

Assuming that you place the `replace` plugin as the first plugin in your setup for Rollup, the `replace` plugin as far as I can tell does not execute the replacement on all files before later plugins or Rollup in general runs. Instead, the execution alternates between doing replacement with the `replace` plugin and then running other plugins/Rollup, at least as far as I can tell.

I would believe that this works well for certain usages of the `replace` plugin, but if you for instance use Typescript, and you attempt to do work with your replacements that require that all replacements are completed before compilation begins (such as renaming fields or variables), you may end up with Typescript stopping compilation with a compilation error after replacements are half-way done.

I don't know whether this is a bug or if it is working as intended. However, if it is not a bug, the lack of documentation reg. this behaviour for the `replace` plugin places an undue burden on users to understand the Rollup lifecycle. I must admit that I also find it surprising that replacements do not occur before all other parts. The documentation itself agrees that replacements should generally be done early on/first:

> Typically, `@rollup/plugin-replace` should be placed in `plugins` _before_ other plugins so that they may apply optimizations, such as dead code removal.

But despite placing the `replace` plugin first, replacements are not performed on all files before other plugins/Rollup executes.

### Your Proposal for Changes

If it is not a bug, it should be documented in https://github.com/rollup/plugins/blob/master/packages/replace/README.md that the `replace` plugin does not perform all replacements before other plugins/Rollup are executed, which may cause issues with compilation if replacement needs to be done before then. If you happen to know of an easy work-around, it would be very nice for this work-around to be documented - disabling type-checking maybe? Though that is not the nicest work-around, since it would be very nice to have type checking, including when doing this kind of renaming. Is there some way to force all replacements for all files/chunks to be done before all other plugins/Rollup executes?

**Comment by MelvinWM** [comment]
(is there a way to add the 'plugin-replace' label to this issue for non-project-members?).

**Comment by TonySpegel** [comment]
I do experience the same behaviour and my plugin section looks like this
```javascript
plugins: [
    replace({
        preventAssignment: true,
        'Reflect.decorate': 'undefined',
        '<!-- ## no-contextmenu -->': `<script>
                window.addEventListener('contextmenu', (e) => {
                    e.preventDefault();
                }, false);
            </script>`,
    }),
    html({
        input: 'dev/index.html',
    }),
    nodeResolve(),
    minifyHTML(),
    terser({
        ecma: 2020,
        module: true,
        warnings: true,
        mangle: {
            properties: {
                regex: /^__/,
            },
        },
    }),
    summary(),
],
```
it doesn't matter where you would put this, it won't replace anything (in my case index.html)

**Comment by MelvinWM** [comment]
> it doesn't matter where you would put this, it won't replace anything (in my case index.html)

Are you certain what you describe is the same as what I describe? In what I describe, replacements do happen, but replacements are not all done before other plugins/steps reg. Rollup, it instead seems to be interleaved, even when putting the `replace` as the first plugin.

**Comment by christian-bromann** [comment]
I am having the same issue. I don't get `replace` to work at all. Sad to see that the issue just stales only because it is not getting triaged 😢
